attest/signer.go

```diff
@@ -0,0 +1,128 @@
+package attest
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"encoding/asn1"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var tpm2HashFuncMap = map[crypto.Hash]tpm2.Algorithm{
+	crypto.SHA1:   tpm2.AlgSHA1,
+	crypto.SHA256: tpm2.AlgSHA256,
+	crypto.SHA384: tpm2.AlgSHA384,
+	crypto.SHA512: tpm2.AlgSHA512,
+}
+
+func tpm2HashFunc(h crypto.Hash) (tpm2.Algorithm, error) {
+	if a, ok := tpm2HashFuncMap[h]; ok {
+		return a, nil
+	}
+	return 0, fmt.Errorf("unsupported hash algorithm 0x%x", h)
+}
+
+type ec20Key struct {
+	tpm    io.ReadWriter
+	handle tpmutil.Handle
+	pub    crypto.PublicKey
+}
+
+type ecdsaSignature struct {
+	R, S *big.Int
+}
+
+func (k *ec20Key) Public() crypto.PublicKey {
+	return k.pub
+}
+
+func (k *ec20Key) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash, err := tpm2HashFunc(opts.HashFunc())
+	if err != nil {
+		return nil, err
+	}
+	scheme := &tpm2.SigScheme{Hash: hash, Alg: tpm2.AlgECDSA}
```

Comment:
TPM 2.0 Commands reference section 20.2.1 says:

```
If the scheme of keyHandle is not TPM_ALG_NULL, then inScheme shall either be the same scheme as
keyHandle or TPM_ALG_NULL. If the sign attribute is not SET in the key referenced by handle then the
TPM shall return TPM_RC_KEY
```

For most AKs (including ours) the Scheme is RSASSA with a SHA256 hash: https://github.com/google/go-attestation/blob/master/attest/tpm.go#L53, so for those keys a NULL scheme or an identical scheme will need to be provided.

The AK template has `tpm2.FlagSignerDefault` set, which includes the sign bit, so we should be fine with regards to key permissions.